bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -83,16 +200,74 @@ struct SourceProtocol<T: AsyncRead + AsyncWrite + Unpin + Send> {
 
     /// Transport to the destination Instance.
     conn: WebSocketStream<T>,
+
+    /// Guest page table dirty bits to restore in the event of a migration
+    /// failure, so that a subsequent migration can attempt to offer only dirty
+    /// pages.
+    ///
+    /// Yes, this representation may seem somewhat weird. Bear with me: when we
+    /// call the `VM_NPT_OPERATION` ioctl, the pages whose dirty bits we would
+    /// like to set are represented by a bitmap. Therefore, we need to be able
+    /// to produce bitmaps for each chunk of guest physical address space for
+    /// which we've previously observed dirty pages. We could, of course, just
+    /// store a map of base guest addresses to 4096-byte arrays, keeping the
+    /// whole bitmap in memory until the migration finishes, but this seems kind
+    /// of sad, given that the bitmap may be fairly sparse.
+    ///
+    /// Instead, we store a map of base guest addresses to a list of indices
```

Comment:
This representation seems overly complex, and moreover, perhaps not as efficient as one might think.

Consider that each tuple in the vec will almost certainly be aligned on a 16-byte boundary, meaning that 256 entries will take more than 4KiB of host memory, including overhead from the Vec. Combined with the overhead of the hash map, I can see some scenarios when this may be slightly more compact, but I can also see many where it's not.  It's really only a win in cases where modifications to the address space are extremely sparse, and it's not clear to me that that's the common case.

I think that, in absence of evidence that memory is unduly affected by storing bitmaps for resetting dirty bits, I'd just use a bitmap.  Recall also that that's an incredibly efficient representation; since each 4KiB page covers 128 MiB of address space.

I'd certainly acquiesce in my objection if there were some measurements that showed it to be a win on average?

Reply:
> this code is assuming that guest address spaces are sparser than they are in real life

Yeah.  Save for guests which are migrated soon after they are provisioned, I would be shocked if any instance which had been up for a day or more (probably hours even) left any page un-dirtied.

So while it will be nice to cut down on migration time/cost for freshly provisioned instances, we probably shouldn't over-rotate on optimizing for that case.

Reply:
This begs the question: is this optimization worth it, then? If the average case is that O(all) pages have been touched in a fairly short time after the VM starts, and we don't expect migrations to happen _that_ frequently, perhaps it's best to just stick with the simpler approach?